spring-batch-infrastructure/src/test/java/org/springframework/batch/item/jms/builder/JmsItemReaderBuilderTests.java

```diff
@@ -0,0 +1,127 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.batch.item.jms.builder;
+
+import java.util.Date;
+
+import javax.jms.Message;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import org.springframework.batch.item.jms.JmsItemReader;
+import org.springframework.jms.core.JmsOperations;
+import org.springframework.jms.core.JmsTemplate;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author Glenn Renfro
+ */
+public class JmsItemReaderBuilderTests {
+
+	private JmsOperations defaultJmsTemplate;
+
+	@Before
+	public void setupJmsTemplate() {
+		this.defaultJmsTemplate = mock(JmsOperations.class);
+		when(this.defaultJmsTemplate.receiveAndConvert()).thenReturn("foo");
+	}
+
+	@Test
+	public void testNoItemType() {
+		JmsItemReader<String> itemReader = new JmsItemReaderBuilder<String>().jmsTemplate(this.defaultJmsTemplate)
+				.build();
+		assertEquals("foo", itemReader.read());
+	}
+
+	@Test
+	public void testSetItemType() {
+		JmsItemReader<String> itemReader = new JmsItemReaderBuilder<String>().jmsTemplate(this.defaultJmsTemplate)
+				.itemType(String.class).build();
+
+		assertEquals("foo", itemReader.read());
+	}
+
+	@Test
+	public void testSetItemSubclassType() {
+		JmsOperations jmsTemplate = mock(JmsOperations.class);
+
+		Date date = new java.sql.Date(0L);
+		when(jmsTemplate.receiveAndConvert()).thenReturn(date);
+
+		JmsItemReader<Date> itemReader = new JmsItemReaderBuilder<Date>().jmsTemplate(jmsTemplate).itemType(Date.class)
+				.build();
+		assertEquals(date, itemReader.read());
+	}
+
+	@Test
+	public void testSetItemTypeMismatch() {
+		JmsItemReader<Date> itemReader = new JmsItemReaderBuilder<Date>().jmsTemplate(this.defaultJmsTemplate)
+				.itemType(Date.class).build();
+		try {
+			itemReader.read();
+			fail("Expected IllegalStateException");
+		}
+		catch (IllegalStateException e) {
+			// expected
+			assertTrue(e.getMessage().indexOf("wrong type") >= 0);
+		}
+	}
+
+	@Test
+	public void testNextMessage() {
+		JmsOperations jmsTemplate = mock(JmsOperations.class);
+		Message message = mock(Message.class);
+		when(jmsTemplate.receive()).thenReturn(message);
+
+		JmsItemReader<Message> itemReader = new JmsItemReaderBuilder<Message>().jmsTemplate(jmsTemplate)
+				.itemType(Message.class).build();
+		assertEquals(message, itemReader.read());
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testTemplateWithNoDefaultDestination() {
+		JmsTemplate jmsTemplate = new JmsTemplate();
+		jmsTemplate.setReceiveTimeout(100L);
+		new JmsItemReaderBuilder<String>().jmsTemplate(jmsTemplate).itemType(String.class).build();
+
+	}
+
+	@Test(expected = IllegalArgumentException.class)
+	public void testTemplateWithNoTimeout() {
```

Comment:
We aren't testing the behavior of the reader itself, we're testing the behavior of the builder.